grove/measurements/estimation.py

```diff
@@ -0,0 +1,161 @@
+    index_mapper = dict(zip(active_qubit_indices,
+                            range(len(active_qubit_indices))))
+
+    results = np.zeros((len(pauli_terms), len(bitstring_results)))
+
+    # convert to array so we can fancy index into it later.
+    # list() is needed to cast because we don't really want a map object
+    bitstring_results = list(map(np.array, bitstring_results))
+    for row_idx, term in enumerate(pauli_terms):
+        memory_index = np.array(list(map(lambda x: index_mapper[x],
+                                         sorted(term.get_qubits()))))
+
+        results[row_idx, :] = [-2 * (sum(x[memory_index]) % 2) + \
+                               1 for x in bitstring_results]
+    return results
+
+
+def estimate_pauli_set(pauli_term_set, basis_transform_dict, program,
+                       variance_bound, quantum_resource):
+    """
+    Estimate the mean of a sum of pauli terms to set variance
+
+    The sample variance is calculated by
+
+    .. math::
+        \begin{align}
+        \mathrm{Var}[\hat{\langle H \rangle}] = \sum_{i, j}h_{i}h_{j}
+        \mathrm{Cov}(\hat{\langle P_{i} \rangle}, \hat{\langle P_{j} \rangle})
+        \end{align}
+
+    :param pauli_term_set: list of pauli terms to measure simultaneously
+    :param basis_transform_dict: basis transform dictionary where the key is
+                                 the qubit index and the value is the basis to
+                                 rotate into. Valid basis is [I, X, Y, Z].
+    :param program: program generating a state to sample from
+    :param variance_bound:  Bound on the variance of the estimator for the
+                            PauliSum. Remember this is the SQUARE of the
+                            standard error!
+
+    :param quantum_resource: quantum abstract machine object
+    :return: estimated expected value, covariance matrix, variance of the
+             estimator, and the number of shots taken
+    """
+    if not isinstance(pauli_term_set, list):
+        raise TypeError("pauli_term_set needs to be a list")
+
+    # check if each term commutes with everything
+    if len(commuting_sets(sum(pauli_term_set))) != 1:
+        raise CommutationError("Not all terms commute in the expected way")
+
+    pauli_for_rotations = PauliTerm.from_list(
+        [(value, key) for key, value in basis_transform_dict.items()])
+    post_rotations = get_rotation_program(pauli_for_rotations)
+
+    coeff_vec = np.array(
+        list(map(lambda x: x.coefficient, pauli_term_set))).reshape((-1, 1))
+
+    results = None
+    sample_variance = np.infty
+    while sample_variance > variance_bound:
+        # note: bit string values are returned according to their listed order
+        # in run_and_measure.  Therefore, sort beforehand to keep alpha numeric
+        tresults = quantum_resource.run_and_measure(
+            program + post_rotations,
+            sorted(list(basis_transform_dict.keys())),
+            trials=10000)
```

Comment:
should this be configurable / adapted to the desired variance bound?

Reply:
I can do that.  The number of samples is bounded assuming all terms have maximal variance.  I will keep the 10000-shot blocks so we don't eagerly saturate the pessimistic bound.